File: backend/src/candidate/candidate.controller.ts
import { Controller, Get, Post, Param, Query, Body } from '@nestjs/common';
import { CandidateService } from './candidate.service';
import { RejectionService } from '../rejection/rejection.service';
import { RejectCandidateDto } from '../rejection/dto/feedback.dto';
import { TargetRole } from '../config/tech-stack.config';

@Controller('candidates')
export class CandidateController {
  constructor(
    private candidateService: CandidateService,
    private rejectionService: RejectionService,
  ) {}

  @Get()
  async findAll(
    @Query('page') page?: string,
    @Query('limit') limit?: string,
    @Query('sortBy') sortBy?: 'totalScore' | 'followers' | 'crawledAt',
    @Query('order') order?: 'asc' | 'desc',
    @Query('search') search?: string,
    @Query('source') source?: string,
    @Query('minScore') minScore?: string,
    @Query('maxScore') maxScore?: string,
    @Query('excludeRejected') excludeRejected?: string,
    @Query('autoExclude') autoExclude?: string,
    @Query('role') role?: TargetRole,
    @Query('recentActivityOnly') recentActivityOnly?: string,
    @Query('activityMonths') activityMonths?: string,
  ) {
    return this.candidateService.findAll({
      page: page ? parseInt(page) : undefined,
      limit: limit ? parseInt(limit) : undefined,
      sortBy,
      order,
      search,
      source,
      minScore: minScore ? parseFloat(minScore) : undefined,
      maxScore: maxScore ? parseFloat(maxScore) : undefined,
      excludeRejected: excludeRejected === 'true',
      autoExclude: autoExclude === 'true',
      role,
      recentActivityOnly: recentActivityOnly === 'true',
      activityMonths: activityMonths ? parseInt(activityMonths) : undefined,
    });
  }

  @Get('stats')
  async getStats() {
    return this.candidateService.getStats();
  }

  @Get('sources')
  async getSources() {
    return this.candidateService.getSources();
  }

  @Get(':id')
  async findOne(@Param('id') id: string) {
    return this.candidateService.findOne(id);
  }

  @Get('username/:username')
  async findByUsername(@Param('username') username: string) {
    return this.candidateService.findByUsername(username);
  }

  @Post(':id/reject')
  async rejectCandidate(
    @Param('id') id: string,
    @Body() dto: RejectCandidateDto,
  ) {
    return this.rejectionService.rejectCandidate(id, dto.reason, dto.notes);
  }

  @Post(':id/shortlist')
  async shortlistCandidate(@Param('id') id: string) {
    return this.rejectionService.shortlistCandidate(id);
  }

  @Post(':id/undo')
  async undoFeedback(@Param('id') id: string) {
    return this.rejectionService.undoFeedback(id);
  }
}

File: backend-java/src/main/java/com/peopleminer/candidate/CandidateController.java
package com.peopleminer.candidate;

import com.peopleminer.config.TechStackConfig.TargetRole;
import com.peopleminer.domain.entity.Candidate;
import com.peopleminer.rejection.RejectionService;
import com.peopleminer.rejection.dto.RejectCandidateRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/candidates")
@RequiredArgsConstructor
public class CandidateController {

    private final CandidateService candidateService;
    private final RejectionService rejectionService;

    @GetMapping
    public ResponseEntity<CandidateListResponse> findAll(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String order,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String source,
            @RequestParam(required = false) Double minScore,
            @RequestParam(required = false) Double maxScore,
            @RequestParam(required = false) Boolean excludeRejected,
            @RequestParam(required = false) Boolean autoExclude,
            @RequestParam(required = false) TargetRole role,
            @RequestParam(required = false) Boolean recentActivityOnly,
            @RequestParam(required = false) Integer activityMonths
    ) {
        CandidateSearchCriteria criteria = CandidateSearchCriteria.builder()
                .page(page)
                .limit(limit)
                .sortBy(sortBy)
                .order(order)
                .search(search)
                .source(source)
                .minScore(minScore)
                .maxScore(maxScore)
                .excludeRejected(excludeRejected)
                .autoExclude(autoExclude)
                .role(role)
                .recentActivityOnly(recentActivityOnly)
                .activityMonths(activityMonths)
                .build();

        return ResponseEntity.ok(candidateService.findAll(criteria));
    }

    @GetMapping("/stats")
    public ResponseEntity<CandidateStatsResponse> getStats() {
        return ResponseEntity.ok(candidateService.getStats());
    }

    @GetMapping("/sources")
    public ResponseEntity<List<SourceCountResponse>> getSources() {
        return ResponseEntity.ok(candidateService.getSources());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Candidate> findById(@PathVariable String id) {
        return candidateService.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/username/{username}")
    public ResponseEntity<Candidate> findByUsername(@PathVariable String username) {
        return candidateService.findByUsername(username)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping("/{id}/reject")
    public ResponseEntity<Map<String, Object>> rejectCandidate(
            @PathVariable String id,
            @RequestBody RejectCandidateRequest request
    ) {
        return ResponseEntity.ok(rejectionService.rejectCandidate(id, request.getReason(), request.getNotes()));
    }

    @PostMapping("/{id}/shortlist")
    public ResponseEntity<Map<String, Object>> shortlistCandidate(@PathVariable String id) {
        return ResponseEntity.ok(rejectionService.shortlistCandidate(id));
    }

    @PostMapping("/{id}/undo")
    public ResponseEntity<Map<String, Object>> undoFeedback(@PathVariable String id) {
        return ResponseEntity.ok(rejectionService.undoFeedback(id));
    }
}
